projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -72,76 +73,55 @@ export const createPrompt = (
-	const [ errorMessage, setErrorMessage ] = useState( false );
-	const { tracks } = useAnalytics();
-
-	// Let's grab post data so that we can do something smart.
-	const currentPostTitle = useSelect( select =>
-		select( 'core/editor' ).getEditedPostAttribute( 'title' )
-	);
-
-	let loading = false;
-	const categories =
-		useSelect( select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ) ) || [];
-
-	const categoryObjects = useSelect(
+function useTaxonomy( postAttribute, taxonomy ) {
+	let loaded = false;
+	const taxonomies =
+		useSelect( select => select( 'core/editor' ).getEditedPostAttribute( postAttribute ) ) || [];
+	const taxonomyObjects = useSelect(
 		select => {
-			return categories
-				.map( categoryId => {
-					const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
+			return taxonomies
+				.map( taxonomyId => {
+					const _taxonomy = select( 'core' ).getEntityRecord( 'taxonomy', taxonomy, taxonomyId );
 
-					if ( ! category ) {
-						// Data is not yet loaded
-						loading = true;
+					if ( ! _taxonomy ) {
+						loaded = true;
 						return;
 					}
 
-					return category;
+					return _taxonomy;
 				} )
 				.filter( Boolean ); // Remove undefined values
 		},
-		[ categories ]
+		[ taxonomies ]
 	);
+	return [ taxonomyObjects, loaded ];
+}
 
-	const tags =
-		useSelect( select => select( 'core/editor' ).getEditedPostAttribute( 'tags' ), [] ) || [];
-
-	const tagObjects = useSelect(
-		select => {
-			return tags
-				.map( tagId => {
-					const tag = select( 'core' ).getEntityRecord( 'taxonomy', 'post_tag', tagId );
-
-					if ( ! tag ) {
-						// Data is not yet loaded
-						loading = true;
-						return;
-					}
+export default function Edit( { attributes: { state }, setAttributes, clientId } ) {
+	const [ content, setContent ] = useState( '' );
+	const [ needsMoreCharacters, setNeedsMoreCharacters ] = useState( false );
+	const [ showRetry, setShowRetry ] = useState( false );
+	const [ errorMessage, setErrorMessage ] = useState( false );
+	const { tracks } = useAnalytics();
 
-					return tag;
-				} )
-				.filter( Boolean ); // Remove undefined values
-		},
-		[ tags ]
+	// Let's grab post data so that we can do something smart.
+	const currentPostTitle = useSelect( select =>
+		select( 'core/editor' ).getEditedPostAttribute( 'title' )
 	);
 
-	useEffect( () => {
-		setIsLoadingCategories( loading );
-	}, [ loading ] );
-
 	const postId = useSelect( select => select( 'core/editor' ).getCurrentPostId() );
+
+	const [ categoryObjects, categoriesLoaded ] = useTaxonomy( 'categories', 'category' );
 	const categoryNames = categoryObjects
 		.filter( cat => cat.id !== 1 )
 		.map( ( { name } ) => name )
 		.join( ', ' );
+
+	const [ tagObjects, tagsLoaded ] = useTaxonomy( 'tags', 'post_tag' );
 	const tagNames = tagObjects.map( ( { name } ) => name ).join( ', ' );
 
+	const isTaxonomyLoaded = categoriesLoaded && tagsLoaded;
```

Comment:
So `isTaxonomyLoaded` never gets to True. I could not debug more, for some reason PHPStorm keeps on freezing....

But the refactor in itself seems like a very good idea (`isLoadingCompletion` is defintiely superfluous now)...